### PR TITLE
Move the torch based ice emulator to soca 

### LINKED ISSUE #1014: ML based balance
## Description
Develop an ML based balance operator as a saber outer block, I think that's what it should be, could you confirm @travissluka and/or @shlyaeva ?

Hard to tell yet in how many steps I will do this work, but the initial goal is to call pre-trained AI balance emulators (developed in the `daml` repo) from within `soca`. I'll add a training option later on.

## Help!
Build `libtorch` as part of the containers and on HPC. Help with this would be nice, the `gnu` version is easy to build, but I couldn't figure out how to build with intel.

### PR DESCRIPTION
**Pull request by guillaumevernieres**
## Description
- Moved some of the AI code from the daml repo to soca. I'm only moving the sea-ice balance, the other balances are not implemented or don't work yet.
- Added a training ctest to generate the weights of the emulator instead of storing pre-computed weights
- switched to using `libtorch` from the spack-stack, but it's unfortunately not installed as a cmake package (can't use find_package)

### Issue(s) addressed
- fixes #1014 

## Testing
**How were these changes tested?**
**Are the changes covered by ctests? (If not, tests should be added first.)**
3 ctests: training, dirac and 3dvar
It's missing a ctest for the adjoint ... 

**What compilers / HPCs was it tested with?**
Only works with `gnu`, I have no idea how to build `torch` with intel, and it's also not in the intel spack-stack
Tested on hercules only

### Arctic
bkg-incr-mlb

### Antarctic
Animated distraction for the reviewer:
incr-sst-icec

### CODE REVIEW

**Comment by guillaumevernieres** [comment]
I broke the `cmake` logic to skip the build of `MLBalance`, oops ... I'll look into it this morning.

**Comment by shlyaeva** [comment]
I pushed a small change: removed linking daml and renamed variables (removed `_category` from the name); ml tests pass on orion for me when passing -DTorch_ROOT during ecbuild step 🎉

File: src/soca/MLBalance/KEmul/BaseEmul.h [hunk 1]
```diff
@@ -0,0 +1,293 @@
...
+    int batchSize_;
+    size_t epochs_;
+    std::string modelOutputFileName_;
+    std::shared_ptr<Net> model_;
+    const eckit::mpi::Comm & comm_;
+    const eckit::Configuration & config_;
+
+   public:
+    // Constructor
+    explicit BaseEmul(const eckit::Configuration & config, const eckit::mpi::Comm & comm):
```
Comment (shlyaeva) [inline src/soca/MLBalance/KEmul/BaseEmul.h]: nitpick: no need for explicit here and elsewhere for ctors with more than one argument.

File: src/soca/MLBalance/KEmul/BaseEmul.h [hunk 2]
```diff
@@ -0,0 +1,293 @@
...
+
+  // -----------------------------------------------------------------------------
+  /// Utilities
+  // -----------------------------------------------------------------------------
+  // -----------------------------------------------------------------------------
+  /// Emulator base class
+  // -----------------------------------------------------------------------------
+  template <typename Net>
+  class BaseEmul {
+   public:
```
Comment (shlyaeva) [inline src/soca/MLBalance/KEmul/BaseEmul.h]: nitpick: perhaps these could be private?

File: src/soca/MLBalance/KEmul/BaseEmul.h [hunk 3]
```diff
@@ -0,0 +1,293 @@
...
+      model_->initWeights();
+
+      // Load model if asked in the config
+      if (config_.has("ffnn.load model")) {
+        std::string modelFileName;
+        config_.get("ffnn.load model", modelFileName);
+        torch::load(model_, modelFileName);
+        model_->loadNorm(modelFileName);
+        /*
+        TODO(G): Move that to oops log debug?
```
Comment (shlyaeva) [inline src/soca/MLBalance/KEmul/BaseEmul.h]: would be nice to move to debug or info or test log here and elsewhere.

File: src/soca/MLBalance/KEmul/BaseEmul.h [hunk 4]
```diff
@@ -0,0 +1,293 @@
...
+      // Track all operations on the tensor to enable automatic differentiation
+      input.set_requires_grad(true);
+
+      // Train the model
+      float finalLoss(0.0);
+      oops::Log::trace() << "Train ..." << std::endl;
+      for (size_t epoch = 0; epoch < epochs_; ++epoch) {
+        // Setup the model for training
+        // TODO(G): figure out why I commented the following line
+        // model_->train();
```
Comment (shlyaeva) [inline src/soca/MLBalance/KEmul/BaseEmul.h]: Perhaps because it has no effect in this case? https://pytorch.org/docs/stable/generated/torch.nn.Module.html#torch.nn.Module.train
(I don't know, just guessing).

File: src/soca/MLBalance/KEmul/BaseEmul.h [hunk 5]
```diff
@@ -0,0 +1,293 @@
...
+        for (int i = 0; i < barWidth; ++i) {
+          if (i < barLength) {
+            std::cout << "=";
+          } else {
+            std::cout << " ";
+          }
+        }
+        std::cout << "] " << std::setw(3) << static_cast<int>(percentage * 100)
+                  << "% "<< "Loss: " << loss << "\r";
+        std::cout.flush();
```
Comment (shlyaeva) [inline src/soca/MLBalance/KEmul/BaseEmul.h]: would be good to move this to an oops info log and use `std::endl` (doesn't render correctly for me on orion otherwise).

File: src/soca/MLBalance/KEmul/IceEmul.h [hunk 6]
```diff
@@ -0,0 +1,232 @@
...
+    prepData(const std::string& fileName, bool geoloc = false, int n = 400000) override {
+      // TODO(G): Move this elsewhere and leverage soca and/or atlas io
+      // Read additional config
+      std::string pole;
+      config_.get("domain.pole", pole);
+      bool cleanData;
+      config_.get("domain.clean data", cleanData);
+
+      // Read the patterns/targets
+      std::vector<float> lat = readCice(fileName, "ULAT");
```
Comment (shlyaeva) [inline src/soca/MLBalance/KEmul/IceEmul.h]: May be nice to move file open/close outside of this routine so it's only opening file once.

**Review by shlyaeva — COMMENTED** [review]
Looks good to me, though I haven't played with changing parameters etc yet.
I have a few minor technical suggestions below and I don't mind making those changes myself if you'd like @guillaumevernieres.

I have one scientific question. Application of just this balance operator (K I K^T) to a dirac on ice concentration leads to reasonably looking increments in other variables, but also an increased ice concentration increment (it's no longer 1, but a value larger than 1). With a simple 2x2 matrix math I see why this happens, but it still goes against my naive intuition of what a balance operator should do. Does anyone understand whether that's desired behavior (and why if so)?

File: src/soca/MLBalance/KEmul/BaseEmul.h (hunk 4 above)
Comment (guillaumevernieres) [reply src/soca/MLBalance/KEmul/BaseEmul.h]: Ha! yes, you're right. I vaguely remember something about normalizing the input/output using a torch generic method  but I couldn't figure out how to make it work. The normalization is done separately, so that line can probably be deleted.

File: src/soca/MLBalance/KEmul/IceEmul.h (hunk 6 above)
Comment (guillaumevernieres) [reply src/soca/MLBalance/KEmul/IceEmul.h]: good catch ... yes. 
Keep in mind that section needs to be re-written. I had access to oops but not soca when I coded this in the `daml` repo. We should leverage the io from soca and eventually add the possibility to read multiple backgrounds.

File: src/soca/MLBalance/KEmul/BaseEmul.h (hunk 2 above)
Comment (guillaumevernieres) [reply src/soca/MLBalance/KEmul/BaseEmul.h]: Damn ... OK. I'll add the get/set functions and switch the data members to private.

File: src/soca/MLBalance/KEmul/BaseEmul.h (hunk 5 above)
Comment (guillaumevernieres) [reply src/soca/MLBalance/KEmul/BaseEmul.h]: I switched to oops::Log , it seems to render properly for me ...

File: CMakeLists.txt [hunk 7]
```diff
@@ -21,7 +21,7 @@ include( soca_compiler_flags )
 ################################################################################
 # Dependencies
 ################################################################################
-find_package( NetCDF REQUIRED COMPONENTS Fortran C)
+find_package( NetCDF REQUIRED COMPONENTS Fortran C CXX)
```
Comment (travissluka) [inline CMakeLists.txt]: @guillaumevernieres The infra team was previously nagging us to get rid of dependency on the C++ bindings of NetCDF (The only remaining repo was soca-science), as it is not well supported by the netcdf people. Please switch to the C bindings instead, I don't want to add the C++ bindings to soca.

**Review by travissluka — CHANGES_REQUESTED** [review]

File: CMakeLists.txt (hunk 7 above)
Comment (guillaumevernieres) [reply CMakeLists.txt]: Done (pending ci status), but it's ugly. It took me a while to realize that the c netcdf binding doesn't deal with degenerate dimension implicitly, ouch.
@travissluka , @shlyaeva , don't get too worked up with the `prepData` function. I'll re-write this from scratch next time I work on this project.

File: src/soca/MLBalance/KEmul/BaseEmul.h [hunk 8]
```diff
@@ -51,11 +51,39 @@ namespace soca {
...
+    void setInputSize(int inputSize) { inputSize_ = inputSize; }
+    void setOutputSize(int outputSize) { outputSize_ = outputSize; }
+    void setHiddenSize(int hiddenSize) { hiddenSize_ = hiddenSize; }
+    void setKernelSize(int kernelSize) { kernelSize_ = kernelSize; }
+    void setStride(int stride) { stride_ = stride; }
+    void setBatchSize(int batchSize) { batchSize_ = batchSize; }
+    void setEpochs(size_t epochs) { epochs_ = epochs; }
+    void setModelOutputFileName(const std::string & modelOutputFileName)
+                               { modelOutputFileName_ = modelOutputFileName; }
+    void setModel(const std::shared_ptr<Net> & model) { model_ = model; }
```
Comment (shlyaeva) [inline src/soca/MLBalance/KEmul/BaseEmul.h]: I don't think you need any of the setters (which is what I was hoping for!).

**Review by shlyaeva — APPROVED** [review]
Thank you for addressing my comments! I agree, some code can be further improved, but OK with me to do that later.

File: src/soca/MLBalance/KEmul/BaseEmul.h (hunk 8 above)
Comment (guillaumevernieres) [reply src/soca/MLBalance/KEmul/BaseEmul.h]: Oops. Good catch. Removed.

**Review by travissluka — APPROVED** [review]

**Comment by shlyaeva** [comment]
> I have one scientific question. Application of just this balance operator (K I K^T) to a dirac on ice concentration leads to reasonably looking increments in other variables, but also an increased ice concentration increment (it's no longer 1, but a value larger than 1). With a simple 2x2 matrix math I see why this happens, but it still goes against my naive intuition of what a balance operator should do. Does anyone understand whether that's desired behavior (and why if so)?

Following up on my question to close this thread. Here, balance operator transforms ice concentration into the balanced and unbalanced parts. The Identity covariance used in the simplified dirac test corresponds to a diagonal matrix with variance of the temperature error and variance of the _unbalanced_ ice concentration error on the diagonal (both equal to 1). Variance of the total ice concentration error (total here == balanced + unbalanced) is higher and not equal to one, so this result is expected. Good explanation of balance operators [for ocean] in [Weaver et al's ECMWF tech memo 491](https://www.ecmwf.int/sites/default/files/elibrary/2006/13007-multivariate-balance-operator-variational-ocean-data-assimilation.pdf).
